examples/text_to_image/train_text_to_image_sdxl.py

```diff
@@ -0,0 +1,1164 @@
+        weight_dtype = torch.bfloat16
+
+    # Move unet, vae and text_encoder to device and cast to weight_dtype
+    # The VAE is in float32 to avoid NaN losses.
+    unet.to(accelerator.device, dtype=weight_dtype)
+    if args.pretrained_vae_model_name_or_path is None:
+        vae.to(accelerator.device, dtype=torch.float32)
+    else:
+        vae.to(accelerator.device, dtype=weight_dtype)
+    text_encoder_one.to(accelerator.device, dtype=weight_dtype)
+    text_encoder_two.to(accelerator.device, dtype=weight_dtype)
+
+    # Create EMA for the unet.
+    if args.use_ema:
+        ema_unet = UNet2DConditionModel.from_pretrained(
+            args.pretrained_model_name_or_path, subfolder="unet", revision=args.revision
+        )
+        ema_unet = EMAModel(ema_unet.parameters(), model_cls=UNet2DConditionModel, model_config=ema_unet.config)
+
+    if args.enable_xformers_memory_efficient_attention:
+        if is_xformers_available():
+            import xformers
+
+            xformers_version = version.parse(xformers.__version__)
+            if xformers_version == version.parse("0.0.16"):
+                logger.warn(
+                    "xFormers 0.0.16 cannot be used for training in some GPUs. If you observe problems during training, please update xFormers to at least 0.0.17. See https://huggingface.co/docs/diffusers/main/en/optimization/xformers for more details."
+                )
+            unet.enable_xformers_memory_efficient_attention()
+        else:
+            raise ValueError("xformers is not available. Make sure it is installed correctly")
+
+    def compute_snr(timesteps):
+        """
+        Computes SNR as per https://github.com/TiankaiHang/Min-SNR-Diffusion-Training/blob/521b624bd70c67cee4bdf49225915f5945a872e3/guided_diffusion/gaussian_diffusion.py#L847-L849
+        """
+        alphas_cumprod = noise_scheduler.alphas_cumprod
+        sqrt_alphas_cumprod = alphas_cumprod**0.5
+        sqrt_one_minus_alphas_cumprod = (1.0 - alphas_cumprod) ** 0.5
+
+        # Expand the tensors.
+        # Adapted from https://github.com/TiankaiHang/Min-SNR-Diffusion-Training/blob/521b624bd70c67cee4bdf49225915f5945a872e3/guided_diffusion/gaussian_diffusion.py#L1026
+        sqrt_alphas_cumprod = sqrt_alphas_cumprod.to(device=timesteps.device)[timesteps].float()
+        while len(sqrt_alphas_cumprod.shape) < len(timesteps.shape):
+            sqrt_alphas_cumprod = sqrt_alphas_cumprod[..., None]
+        alpha = sqrt_alphas_cumprod.expand(timesteps.shape)
+
+        sqrt_one_minus_alphas_cumprod = sqrt_one_minus_alphas_cumprod.to(device=timesteps.device)[timesteps].float()
+        while len(sqrt_one_minus_alphas_cumprod.shape) < len(timesteps.shape):
+            sqrt_one_minus_alphas_cumprod = sqrt_one_minus_alphas_cumprod[..., None]
+        sigma = sqrt_one_minus_alphas_cumprod.expand(timesteps.shape)
+
+        # Compute SNR.
+        snr = (alpha / sigma) ** 2
+        return snr
```

Comment:
isn't this SNR gamma? it causes a divide by zero, as the sigmas are the divisor and when using rescaled betas with zero SNR, well, the last timestep is a zero. and it blows up :D

Reply:
Yeah. What do you suggest? Should we disable Min-SNR?

Reply:
i can state with certainty that simply adding a little bit of noise to make it non-zero doesn't work, but i've pinged someone who may have better insight

Reply:
@sayakpaul maybe just catch the condition where sigma = 0 and emit an exception for now.

Reply:
That will be a bit too brutal no?

Reply:
@sayakpaul allow me to present you the two options. you can pick less brutal :) 

* user doesn't get warning up front. training starts, and once it hits zero sigma, the training goes into NaN state and then the user doesn't have any idea that it happened until they notice black images during validations
* user gets warning up front that the options are incompatible with rescaled_betas_zero_snr and then they can adjust settings accordingly

Reply:
> user doesn't get warning up front. training starts, and once it hits zero sigma, the training goes into NaN state and then the user doesn't have any idea that it happened until they notice black images during validations

I am myself the victim 🥲

> user gets warning up front that the options are incompatible with rescaled_betas_zero_snr and then they can adjust settings accordingly

What can we suggest to them about it? When `snr_gamma` is non-zero and the scheduler config has the option to rescale betas, we throw an error?

Reply:
i would recommend they not use SNR fixes at all in combination with terminal SNR, which seems to encompass most / all of the former solutions.

those other options really seem to work best when using noise_offset parameter.

your idea sounds correct

Reply:
Does the following check work for you?

```python
if "rescale_betas_zero_snr" in scheduler.config:
    if scheduler.config.rescale_betas_zero_snr and args.snr_gamma:
        logger.warn("...")
```

Reply:
if we don't need to `hasattr` because it is always there after v0.18 or whichever release included that, then yeah that kind of works for me. i would still throw an exception and have an arg to override it.

Reply:
Feel free to directly put suggestions (as suggestion blocks) on the code.

Reply:
either i never go to the right location for that in Github, or I don't have permissions to do so.

here is my proposed changes:
```py
    parser.add_argument(
        "--force_snr_gamma",
        action="store_true",
        help=(
            "When using SNR gamma with rescaled betas for zero terminal SNR, a divide-by-zero error can cause NaN"
            " condition when computing the SNR with a sigma value of zero. This parameter overrides the check,"
            " allowing the use of SNR gamma with a terminal SNR model. Use with caution, and closely monitor results."
        )
    )
```


```py
    # Load scheduler and models
    noise_scheduler = DDPMScheduler.from_pretrained(args.pretrained_model_name_or_path, subfolder="scheduler")
    # Check for terminal SNR in combination with SNR Gamma:
    if args.snr_gamma and not args.force_snr_gamma and (hasattr(noise_scheduler.config, 'rescale_betas_zero_snr') and noise_scheduler.config.rescale_betas_zero_snr):
        raise ValueError(
            'The selected noise scheduler for the model {args.pretrained_model_name_or_path} uses rescaled betas for zero SNR.\n'
            'When this configuration is present, the parameter --snr_gamma may not be used without parameter force_snr_gamma.\n'
            'This is due to a mathematical incompatibility between our current SNR gamma implementation, and a sigma value of zero.'
        )
```

Reply:
Done in https://github.com/huggingface/diffusers/pull/4505/commits/41a2580ec84eb792c0a7e8e0a29a49a34e87ec1b 

Thanks for your suggestions.